cmd/terminal.go

```diff
@@ -0,0 +1,333 @@
+		scheme = wsProtocol
+	}
+	return scheme
+}
+
+// TerminalCmd handles the terminal command
+type TerminalCmd struct {
+	server     string
+	skipVerify bool
+	deviceID   string
+	sessionID  string
+	running    bool
+	stop       chan bool
+}
+
+// NewTerminalCmd returns a new TerminalCmd
+func NewTerminalCmd(cmd *cobra.Command, args []string) (*TerminalCmd, error) {
+	server, err := cmd.Flags().GetString(argRootServer)
+	if err != nil {
+		return nil, err
+	}
+
+	skipVerify, err := cmd.Flags().GetBool(argRootSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TerminalCmd{
+		server:     server,
+		skipVerify: skipVerify,
+		deviceID:   args[0],
+		stop:       make(chan bool),
+	}, nil
+}
+
+// Run executes the command
+func (c *TerminalCmd) Run() error {
+	log.Info(fmt.Sprintf("Connecting to the remote terminal of the device %s...", c.deviceID))
+
+	tokenPath, err := getDefaultAuthTokenPath()
+	if err != nil {
+		return errors.Wrap(err, "Unable to determine the auth token path")
+	}
+
+	token, err := ioutil.ReadFile(tokenPath)
+	if err != nil {
+		return errors.Wrap(err, "Please Login first")
+	}
+
+	// connect to the websocket
+	deviceConnectPath := "/api/management/v1/deviceconnect/devices/" + c.deviceID + "/connect"
+	parsedURL, err := url.Parse(c.server)
+	if err != nil {
+		return err
+	}
+
+	scheme := getWebSocketScheme(parsedURL.Scheme)
+	u := url.URL{Scheme: scheme, Host: parsedURL.Host, Path: deviceConnectPath}
+
+	headers := http.Header{}
+	headers.Set("Authorization", "Bearer "+string(token))
+	websocket.DefaultDialer.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: c.skipVerify,
+	}
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), headers)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	// ping-pong
+	conn.SetReadLimit(maxMessageSize)
+	_ = conn.SetReadDeadline(time.Now().Add(defaultPingWait))
+	conn.SetPingHandler(func(message string) error {
+		pongWait, _ := strconv.Atoi(message)
+		_ = conn.SetReadDeadline(time.Now().Add(time.Duration(pongWait) * time.Second))
+		return conn.WriteControl(websocket.PongMessage, []byte{}, time.Now().Add(writeWait))
+	})
+
+	log.Info("Press CTRL+] to quit the session")
+
+	// set the terminal in raw mode
+	oldState, err := term.MakeRaw(0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = term.Restore(0, oldState)
+	}()
+
+	// get the terminal width and height
+	termID := int(os.Stdout.Fd())
+	termWidth, termHeight, err := terminal.GetSize(termID)
+	if err != nil {
+		return err
+	}
+
+	// send the shell start message
+	m := &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:   ws.ProtoTypeShell,
+			MsgType: wsshell.MessageTypeSpawnShell,
+			Properties: map[string]interface{}{
+				"terminal_width":  termWidth,
+				"terminal_height": termHeight,
+			},
+		},
+	}
+
+	data, err := msgpack.Marshal(m)
+	if err != nil {
+		return err
+	}
+	_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+	_ = conn.WriteMessage(websocket.BinaryMessage, data)
```

Comment:
All the above is related to something like `setupTerminal`, no?

Reply:
I'm not sure it makes sense to split it to an independent function `setupTerminal`, as we won't use it elsewhere.
What's your idea, here?

Reply:
Then there's no need :)